advanced_hw_2.py

```diff
@@ -0,0 +1,412 @@
+# -*- coding: utf-8 -*-
+"""Копия блокнота "HW_2.ipynb"
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1euEn6zYDb60UDX9sO0eKnfLLIBUBQVXD
+
+# Задание 1 (2 балла)
+
+Напишите класс `MyDict`, который будет полностью повторять поведение обычного словаря, за исключением того, что при итерации мы должны получать и ключи, и значения.
+
+**Модули использовать нельзя**
+"""
+
+class MyDict(dict):
+    def __iter__(self):
+        for key in self.keys():
+            yield (key, self[key])
+
+dct = MyDict({"a": 1, "b": 2, "c": 3, "d": 25})
+for key, value in dct:
+    print(key, value)
+
+for key, value in dct.items():
+    print(key, value)
+
+for key in dct.keys():
+    print(key)
+
+dct["c"] + dct["d"]
+
+"""# Задание 2 (2 балла)
+
+Напишите функцию `iter_append`, которая "добавляет" новый элемент в конец итератора, возвращая итератор, который включает изначальные элементы и новый элемент. Итерироваться по итератору внутри функции нельзя, то есть вот такая штука не принимается
+```python
+def iter_append(iterator, item):
+    lst = list(iterator) + [item]
+    return iter(lst)
+```
+
+**Модули использовать нельзя**
+"""
+
+def iter_append(iterator, item):
+    yield from iterator
+    yield item 
+    pass
+
+my_iterator = iter([1, 2, 3])
+new_iterator = iter_append(my_iterator, 4)
+
+for element in new_iterator:
+    print(element)
+
+"""# Задание 3 (5 баллов)
+
+Представим, что мы установили себе некотурую библиотеку, которая содержит в себе два класса `MyString` и `MySet`, которые являются наследниками `str` и `set`, но также несут и дополнительные методы.
+
+Проблема заключается в том, что библиотеку писали не очень аккуратные люди, поэтому получилось так, что некоторые методы возвращают не тот тип данных, который мы ожидаем. Например, `MyString().reverse()` возвращает объект класса `str`, хотя логичнее было бы ожидать объект класса `MyString`.
+
+Найдите и реализуйте удобный способ сделать так, чтобы подобные методы возвращали экземпляр текущего класса, а не родительского. При этом **код методов изменять нельзя**
+
+**+3 дополнительных балла** за реализацию того, чтобы **унаследованные от `str` и `set` методы** также возвращали объект интересующего нас класса (то есть `MyString.replace(..., ...)` должен возвращать `MyString`). **Переопределять методы нельзя**
+
+**Модули использовать нельзя**
+"""
+
+def return_other_type(method):
+    def wrapper(*args, **kwargs):
+        result = method(*args, **kwargs)
+        if isinstance(result, str):
+            return args[0].__class__(result)
+        elif isinstance(result, set):
+            return args[0].__class__(result)
```

Comment:
1. Хардкодить типы это не лучшее решение. Декоратор можно сделать универсальным, но вешать его только на те методы, которые мы хотим поменять
2. Если вместо дандера можно использовать обычную функцию, то лучше сделать именно так
```suggestion
        return type(args[0])(result)
```